vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Collections.Generic
+open System.Collections.Immutable
+open System.Composition
+
+open Microsoft.CodeAnalysis.CodeFixes
+open Microsoft.CodeAnalysis.ExternalAccess.FSharp.Diagnostics
+open Microsoft.CodeAnalysis.Text
+
+open CancellableTasks
+
+[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.RemoveUnnecessaryParentheses); Shared; Sealed>]
+type internal FSharpRemoveUnnecessaryParenthesesCodeFixProvider [<ImportingConstructor>] () =
+    inherit CodeFixProvider()
+
+    static let title = SR.RemoveUnnecessaryParentheses()
+    static let fixableDiagnosticIds = ImmutableArray.Create "IDE0047" // TODO: FSharpIDEDiagnosticIds.RemoveUnnecessaryParentheses
+
+    /// IDE0047: Remove unnecessary parentheses.
+    ///
+    /// https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0047-ide0048
+    override _.FixableDiagnosticIds = fixableDiagnosticIds
+
+    override this.RegisterCodeFixesAsync context = context.RegisterFsharpFix this
+
+    override this.GetFixAllProvider() =
+        this.RegisterFsharpFixAll(fun diagnostics ->
+            // There may be pairs of diagnostics with nested spans
+            // for which it would be valid to apply either but not both.
+            let builder = ImmutableArray.CreateBuilder diagnostics.Length
+
+            let spans =
+                SortedSet
+                    { new IComparer<TextSpan> with
+                        member _.Compare(x, y) =
+                            if x.IntersectsWith y then 0 else x.CompareTo y
+                    }
+
+            for i in 0 .. diagnostics.Length - 1 do
+                let diagnostic = diagnostics[i]
+
+                if spans.Add diagnostic.Location.SourceSpan then
+                    builder.Add diagnostic
+
+            builder.ToImmutable())
+
+    interface IFSharpCodeFixProvider with
+        member _.GetCodeFixIfAppliesAsync context =
+            assert (context.Span.Length >= 3) // (…)
+
+            cancellableTask {
+                let! sourceText = context.Document.GetTextAsync context.CancellationToken
+                let txt = sourceText.ToString(TextSpan(context.Span.Start, context.Span.Length))
+
+                let firstChar = txt[0]
+                let lastChar = txt[txt.Length - 1]
+
+                match firstChar, lastChar with
+                | '(', ')' ->
+                    let inline toPat f x = if f x then Some() else None
+                    let (|LetterOrDigit|_|) = toPat Char.IsLetterOrDigit
+                    let (|Punctuation|_|) = toPat Char.IsPunctuation
+                    let (|Symbol|_|) = toPat Char.IsSymbol
```

Comment:
These probably should be outside the state machine, so we're not codegen more than necessary

Reply:
Hmm, I believe they're [compiled outside of the state machine](https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAWAChjgFsUwUAKCFEFHAJwEskBzAShQF5iUBKbAEMIAaxQBvfoNlkKbYGwRCYABWEUqAD14oWVHUIAWydDDwrq3BMAgqAcjCQIZsgfJpQAMuVxMAeSYAERZ2FmwoAH0obh5VDQoAYWNhJgA6AEkIX2x/INDw7Dd3T2ooNQBXJABjbErNFmdo2L1sdU0UFLSsiCra+sbnErlyL0w8ACMYYBa4hM7ujOyJ6eARwQ2BPE0a40YAbQAGAF0UAHcI4y2UKBRc/JCwiJQAWgA+FCZySqZUAEYbnd+nUGtgmqgPl8fn8UAAmIHoDBTGZvT7feqwgDMiKiaOhmNQABYSgBfIA=) anyway. I could also just `inline` them.

Reply:
aa04c18

Reply:
Yeah, generally rule of thumb - the less bindings you have inside the state machine instance - the less codegen, lifting and magic it needs to do to statically compile all the things.